File: library/src/main/java/com/yausername/youtubedl_android/StreamProcessExtractor.java
package com.yausername.youtubedl_android;

import android.util.Log;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.nio.charset.StandardCharsets;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class StreamProcessExtractor extends Thread {

    private static final String TAG = StreamProcessExtractor.class.getSimpleName();

    private static final long ETA = -1;
    private static final float PERCENT = -1.0f;
    private static final int GROUP_PERCENT = 1;
    private static final int GROUP_MINUTES = 2;
    private static final int GROUP_SECONDS = 3;
    private final InputStream stream;
    private final StringBuffer buffer;
    private final DownloadProgressCallback callback;

    private final Pattern p = Pattern.compile("\\[download\\]\\s+(\\d+\\.\\d)% .* ETA (\\d+):(\\d+)");
    private final Pattern pAria2c = Pattern.compile("\\[#\\w{6}.*\\((\\d*\\.*\\d+)%\\).*?((\\d+)m)*((\\d+)s)*]");
    private float progress = PERCENT;
    private long eta = ETA;

    public StreamProcessExtractor(StringBuffer buffer, InputStream stream, DownloadProgressCallback callback) {
        this.stream = stream;
        this.buffer = buffer;
        this.callback = callback;
        this.start();
    }

    public void run() {
        try {
            Reader in = new InputStreamReader(stream, StandardCharsets.UTF_8);
            StringBuilder currentLine = new StringBuilder();
            int nextChar;
            while ((nextChar = in.read()) != -1) {
                buffer.append((char) nextChar);
                if (nextChar == '\r' || nextChar == '\n' && callback != null) {
                    final String line = currentLine.toString();
                    if (line.startsWith("["))
                        processOutputLine(line);
                    currentLine.setLength(0);
                    continue;
                }
                currentLine.append((char) nextChar);
            }
        } catch (final IOException e) {
            if (BuildConfig.DEBUG)
                Log.e(TAG, "failed to read stream", e);
        }
    }

    private void processOutputLine(String line) {
        callback.onProgressUpdate(getProgress(line), getEta(line), line);
    }

    private float getProgress(String line) {
        final Matcher matcher = p.matcher(line);
        if (matcher.find())
            return progress = Float.parseFloat(matcher.group(GROUP_PERCENT));
        else {
            final Matcher mAria2c = pAria2c.matcher(line);
            if (mAria2c.find())
                return progress = Float.parseFloat(mAria2c.group(1));
        }
        return progress;
    }

    private long getEta(String line) {
        final Matcher matcher = p.matcher(line);
        if (matcher.find())
            return eta = convertToSeconds(matcher.group(GROUP_MINUTES), matcher.group(GROUP_SECONDS));
        else {
            final Matcher mAria2c = pAria2c.matcher(line);
            if (mAria2c.find())
                return eta = convertToSeconds(mAria2c.group(3), mAria2c.group(5));
        }
        return eta;
    }

    private int convertToSeconds(String minutes, String seconds) {
        if (seconds == null)
            return 0;
        else if (minutes == null)
            return Integer.parseInt(seconds);
        return Integer.parseInt(minutes) * 60 + Integer.parseInt(seconds);
    }

}

File: library/src/main/java/com/yausername/youtubedl_android/StreamProcessExtractor.kt
package com.yausername.youtubedl_android

import android.util.Log
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.io.Reader
import java.nio.charset.StandardCharsets
import java.util.regex.Pattern

internal class StreamProcessExtractor(
    private val buffer: StringBuffer,
    private val stream: InputStream,
    private val callback: ((Float, Long, String) -> Unit)?
) : Thread() {
    private val p = Pattern.compile("\\[download\\]\\s+(\\d+\\.\\d)% .* ETA (\\d+):(\\d+)")
    private val pAria2c =
        Pattern.compile("\\[#\\w{6}.*\\((\\d*\\.*\\d+)%\\).*?((\\d+)m)*((\\d+)s)*]")
    private var progress = PERCENT
    private var eta = ETA

    init {
        start()
    }

    override fun run() {
        try {
            val input: Reader = InputStreamReader(stream, StandardCharsets.UTF_8)
            val currentLine = StringBuilder()
            val maxLineLength = 10_000  // Limit to 10,000 characters per line
            val maxBufferLength = 1_000_000  // Limit to 1 million characters in total buffer

            var nextChar: Int
            while (input.read().also { nextChar = it } != -1) {

                val c = nextChar.toChar()

                // Safely append to buffer
                synchronized(buffer) {
                    buffer.append(c)
                    if (buffer.length > maxBufferLength) {
                        buffer.delete(0, buffer.length / 2) // truncate older half
                    }
                }

                // Process line if newline character is detected
                if (c == '\r' || c == '\n') {
                    if (callback != null) {
                        val line = currentLine.toString()
                        processOutputLine(line)
                    }
                    currentLine.setLength(0)
                } else {
                    // Protect against excessively long lines
                    if (currentLine.length < maxLineLength) {
                        currentLine.append(c)
                    } else {
                        // Drop the line if it's too long (could also log or throw)
                        currentLine.setLength(0)
                    }
                }
            }
        } catch (e: IOException) {
            if (BuildConfig.DEBUG) Log.e(TAG, "failed to read stream", e)
        } finally {
            try {
                stream.close()
            } catch (ignored: IOException) {
            }
        }
    }

    private fun processOutputLine(line: String) {
        callback?.let { it(getProgress(line), getEta(line), line) }
    }

    private fun getProgress(line: String): Float {
        val matcher = p.matcher(line)
        if (matcher.find()) return matcher.group(GROUP_PERCENT).toFloat()
            .also { progress = it } else {
            val mAria2c = pAria2c.matcher(line)
            if (mAria2c.find()) return mAria2c.group(1).toFloat().also { progress = it }
        }
        return progress
    }

    private fun getEta(line: String): Long {
        val matcher = p.matcher(line)
        if (matcher.find()) return convertToSeconds(
            matcher.group(GROUP_MINUTES),
            matcher.group(GROUP_SECONDS)
        ).also { eta = it.toLong() }.toLong() else {
            val mAria2c = pAria2c.matcher(line)
            if (mAria2c.find()) return convertToSeconds(
                mAria2c.group(3),
                mAria2c.group(5)
            ).also { eta = it.toLong() }.toLong()
        }
        return eta
    }

    private fun convertToSeconds(minutes: String?, seconds: String?): Int {
        if (seconds == null) return 0 else if (minutes == null) return seconds.toInt()
        return minutes.toInt() * 60 + seconds.toInt()
    }

    companion object {
        private val TAG = StreamProcessExtractor::class.java.simpleName
        private const val ETA: Long = -1
        private const val PERCENT = -1.0f
        private const val GROUP_PERCENT = 1
        private const val GROUP_MINUTES = 2
        private const val GROUP_SECONDS = 3
    }
}